### PR TITLE
add docs for interpretation of length scales in periodic kernel

### LINKED ISSUE #3987: GP Periodic Kernel uses uncommon formula
As discussed [on the Discourse](https://discourse.pymc.io/t/gp-periodic-kernel-is-pymc3-different-from-other-sources/5348/2), PyMC3's `Periodic` kernel is [implemented as](https://docs.pymc.io/api/gp/cov.html#pymc3.gp.cov.Periodic):

equation

But, in other sources such as David Duvenaud’s [Kernel Cookbook](https://www.cs.toronto.edu/~duvenaud/cookbook/), as well as the [NB for mean and cov functions](https://docs.pymc.io/notebooks/GP-MeansAndCovs.html?highlight=periodic#Constructing-Periodic-using-WarpedInput), the formula is given by:
equation

According to @bwengals, this changes the interpretation of the lengthscale: a constant factor changes how you think of the lengthscale and what priors you’d set on it. 
So, **ideally, PyMC3's implementation should be changed to the second equation above** -- note that this would not be backwards-compatible then, so it should be accompanied by a warning. _Does anybody disagree?_

A final note: it seems like we’re using another slightly different implementation in PyMC**4**:

equation
@tirthasheshpatel do you confirm it is equivalent to the standard formula? I notice there is a new `sigma ** 2` parameter and the distance between `x` and `x'` is squared 🤔

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
Fixes #3987.

I have added a `versionchanged` just so that the documentation warns the user when the next version is released. Changed the tests accordingly.

cc @bwengals

### CODE REVIEW

**Review by AlexAndorra — APPROVED** [review]
LGTM, thanks @tirthasheshpatel ! I'll wait for @bwengals's approval and merge after that 😉

**Comment by bwengals** [comment]
I think we ought to consider that this may make peoples past models / inference results wrong, and introduce bugs on the users end, especially since informative priors on lengthscales are a good idea, folks might have an informative prior on ell that would need to be re-scaled.   I think here it's not so much a bug, as not using the most common convention.  No users code is wrong or broken right now that uses Periodic (at least not because of this!).

So if someone does rerun their model, or are using a prefitted model in production, their new results after this change will have a very subtle bug now, that will be very tough on their end to figure out.  I don't think we want this outcome.

Either way, we absolutely need to change the docs / docstring to be correct.  But I'm not 100% sure on changing that factor at this point (even if it would have been better from the start).

Thoughts?

**Comment by AlexAndorra** [comment]
Thanks for the detailed answer @bwengals. I agree with this and I think we have two options:

**1. Status-quo**: do nothing and not use the standard Periodic kernel formula. This way, there is no backwards issues but it might be confusing to new users: they _think_ we're using the common kernel, set their priors and lenghtscale according to this parametrization and it turns out they should have scaled their priors otherwise (in short, the inverse problem of what you laid out).

**2. Changing to the common kernel**, with the issues you laid out. I tend to favor this option as it enforces good practices going forward (if I understood correctly what you were saying on the Discourse), which is something we want. From a "rolling out" perspective, this doesn't seem _that_ challenging to me: it's basically akin to a bug whose fixes imply some backwards incompatibility -- but we have experience on how to do that.
If we take this road, maybe we should condition this fix for versions >= 3.9.5 and display a `DeprecationWarning` during 3.9.3 and 3.9.4.

How does it sound?

**Comment by tirthasheshpatel** [comment]
Agreed with both of you! What do say about a `DeprecationWarning` @bwengals. I am cool with keeping it as is or just adding a `DeprecationWarning`

**Comment by junpenglao** [comment]
To me it is a breaking change with little gain. I think we should keep it as is and clarify in the comment of the differences in scaling compare to standard/other definitions.

**Comment by tirthasheshpatel** [comment]
Yeah, I will add a docstring stating the difference between the two conventions.

**Comment by bwengals** [comment]
Ya agree with @junpenglao, best to just fix the docs so it's clear

**Comment by AlexAndorra** [comment]
Ok, so let's go with that @tirthasheshpatel ! I think you can do that directly in this PR 😉

**Comment by tirthasheshpatel** [comment]
Sorry for the late response. Added the docs!

File: pymc3/gp/cov.py
Comment (junpenglao) [inline pymc3/gp/cov.py]:
```suggestion
    The scaling factor for this kernel is different compare to the other standard
    definition (see [1]_), thus change the interpretation of length-scale in a subtle
    way. Please keep that in mind when you are defining priors on length scale,
    especially when you are trying to replicate other results.
    
    You can divided the length-scale by 2 in the kernel initialization to recover the
    standard definition.
```

@bwengals is this correct?

Comment (bwengals) [reply pymc3/gp/cov.py]: If I read that it's different, how should one interpret it?  I think it'd be good to be specific.  

Also small wording suggestion, how does this sound?

Note that the scaling factor for this kernel is different compare**d** to the more common definition (see [1]_).  Here, 0.5 is in the exponent instead of the more common value, 4.  Divide the length-scale by 2 when initializing the kernel to recover the standard definition.

Comment (tirthasheshpatel) [reply pymc3/gp/cov.py]: Sorry for not replying earlier. I missed @bwengals's reply.

> Here, 0.5 is in the exponent instead of the more common value, 4.

I think what you meant here was:

Here, 0.5 is in the exponent instead of the more common value, 2.

Right?

Comment (bwengals) [reply pymc3/gp/cov.py]: whoops, yep you're right!

**Comment by bwengals** [comment]
good to merge?

**Review by AlexAndorra — APPROVED** [review]
Yep, all good now, thanks @tirthasheshpatel and @bwengals !

**Pull request merged**
